JSDemos/Demos/Toolbar/Multiline/React/App.js

```diff
@@ -0,0 +1,335 @@
+  },
+};
+
+const addButtonOptions = {
+  icon: 'add',
+  text: 'Add',
+  onClick: () => {
+    notify('Add button has been clicked!');
+  },
+};
+
+const removeButtonOptions = {
+  icon: 'trash',
+  text: 'Remove',
+  onClick: () => {
+    notify('Remove button has been clicked!');
+  },
+};
+
+const aboutButtonOptions = {
+  icon: 'help',
+  text: 'About',
+  onClick: () => {
+    notify('About button has been clicked!');
+  },
+};
+
+function App() {
+  const [lineHeight, setLineHeight] = React.useState(lineHeights[1].lineHeight);
+  const [fontSize, setFontSize] = React.useState(fontSizes[2].size);
+  const [textAlign, setTextAlign] = React.useState([textAligns[0].alignment]);
+  const [toolbarLineMode, setToolbarLineMode] = React.useState(
+    toolbarLineModes[0],
+  );
+  const [multiline, setMultiline] = React.useState(true);
+  const [checkBoxValue, setCheckBoxValue] = React.useState(false);
+
+  const onCheckBoxChange = React.useCallback(({ value }) => {
+    setCheckBoxValue(value);
+    notify('Checkbox option has been changed!');
+  }, []);
+
+  const onLineHeightChange = React.useCallback(
+    (e) => setLineHeight(e.item.lineHeight),
+    [],
+  );
+
+  const onFontSizeChange = React.useCallback(
+    (e) => setFontSize(e.item.size),
+    [],
+  );
+
+  const onTextAlignChange = React.useCallback((e) => {
+    setTextAlign([e.itemData.alignment]);
+  }, []);
+
+  const onToolbarLineModeChange = React.useCallback(({ value }) => {
+    setMultiline(value === toolbarLineModes[0]);
+    setToolbarLineMode(value);
+  }, []);
+
+  const renderFontSize = React.useCallback(
+    (itemData) => (
+      <div style={{ fontSize: `${itemData.size}px` }}>{itemData.text}</div>
+    ),
+    [],
+  );
+
+  const renderTextAlign = React.useCallback(() => {
+    const isMenu = false;
+
+    return (
+      <ButtonGroup
+        items={getTextAlignItems(isMenu)}
+        keyExpr="alignment"
+        stylingMode="outlined"
+        selectedItemKeys={textAlign}
+        onItemClick={onTextAlignChange}
+      ></ButtonGroup>
+    );
+  }, [textAlign]);
+
+  const renderTextAlignMenu = React.useCallback(() => {
+    const isMenu = true;
+
+    return (
+      <ButtonGroup
+        items={getTextAlignItems(isMenu)}
+        displayExpr="text"
+        keyExpr="alignment"
+        stylingMode="outlined"
+        selectedItemKeys={textAlign}
+        onItemClick={onTextAlignChange}
+      ></ButtonGroup>
+    );
+  }, [textAlign]);
+
+  const renderMenuSeparator = React.useCallback(() => <div className="toolbar-menu-separator"></div>, []);
+
+  return (
+    <React.Fragment>
+      <Resizable
+        className="resizable-container"
+        minWidth={500}
+        minHeight={150}
+        maxHeight={370}
+        handles="right"
+      >
+        <Toolbar multiline={multiline}>
+          <Item location="before">
+            <Button
+              icon="undo"
+              onClick={() => {
+                notify('Undo button has been clicked!');
```

Comment:
all such callbacks should be memoized as well

Reply:
Can you check this commit? https://github.com/DevExpress/devextreme-demos/pull/2004/commits/f74faf018932a9e9b839ac3cf9fef55b51c3e0e9